crates/core/src/host/wasm_common/module_host_actor.rs

```diff
@@ -537,29 +556,56 @@ impl<T: WasmInstance> WasmInstanceActor<T> {
                 .collect();
 
             let mut new_tables = Vec::new();
+            let mut index_changes = Vec::new();
+
             for table in self.info.catalog.values().filter_map(EntityDef::as_table) {
-                let mut proposed_schema = self.schema_for(table)?;
+                let proposed_schema_def = self.schema_for(table)?;
+
                 if let Some(known_schema) = known_tables.remove(&table.name) {
-                    // If the table is known, we also know its id. Update the
-                    // index definitions so the `TableDef` of both schemas is
-                    // equivalent.
-                    for index in proposed_schema.indexes.iter_mut() {
-                        index.table_id = known_schema.table_id;
-                    }
-                    let known_schema = TableDef::from(known_schema);
-                    if known_schema != proposed_schema {
+                    let table_id = known_schema.table_id;
+                    let known_schema_def = TableDef::from(known_schema.clone());
+
+                    if Equiv(&known_schema_def) != Equiv(&proposed_schema_def) {
                         self.system_logger()
                             .warn(&format!("stored and proposed schema of `{}` differ", table.name));
+                        log::warn!("schema mismatch {}", table.name);
                         tainted.push(table.name.to_owned());
                     } else {
-                        // Table unchanged
+                        // Schema unchanged, try to migrate indexes.
+                        let mut known_indexes = known_schema
+                            .indexes
+                            .into_iter()
+                            .map(|idx| (idx.index_name.clone(), idx))
+                            .collect::<BTreeMap<_, _>>();
+
+                        for mut index_def in proposed_schema_def.indexes {
+                            // This is zero in the proposed schema, as the table
+                            // id is not known at proposal time.
+                            index_def.table_id = table_id;
+
+                            match known_indexes.remove(&index_def.name) {
+                                None => index_changes.push((None, Some(index_def))),
+                                Some(known_index) => {
+                                    let known_id = IndexId(known_index.index_id);
+                                    let known_index_def = IndexDef::from(known_index);
+                                    if known_index_def != index_def {
+                                        index_changes.push((Some(known_id), Some(index_def)));
+                                    }
+                                }
+                            }
+                        }
+
+                        // Indexes not in the proposed schema shall be dropped.
+                        for index in known_indexes.into_values() {
```

Comment:
It does seem like `RelationalDB::get_all_tables` won't correctly return all indexes, which means we do not actually drop indexes not present in the new schema.

Reply:
Actually... I think this might have to do with caching of the schema

Reply:
Here we go:

https://github.com/clockworklabs/SpacetimeDB/blob/c010bfadcac34df4860fb9b8ff0cd1d0c548014a/crates/core/src/db/datastore/locking_tx_datastore/mod.rs#L123-L130

This is called when a transaction commits, but the passed `&schema` may have been modified in the transaction.

Reply:
Yes, this is completely possible. I believe when I wrote this I did so under the assumption that schemas wouldn't be changed except in the first transaction or something to that effect.